elf_reader.go

```diff
@@ -585,6 +587,40 @@ func (ec *elfCode) relocateInstruction(ins *asm.Instruction, rel elf.Symbol) err
 			return fmt.Errorf("asm relocation: %s: unsupported type %s", name, typ)
 		}
 
+		// If a Call instruction is found and the datasec has a btf.Func with a Name
+		// that matches the symbol name we mark the instruction as a call to a kfunc.
+		if ins.OpCode.JumpOp() == asm.Call {
+			if ec.btf == nil {
+				break
+			}
+			var ds *btf.Datasec
+			if err := ec.btf.TypeByName(".ksyms", &ds); err != nil {
+				break
+			}
+
+			var found bool
+			for _, v := range ds.Vars {
+
+				// if func check if name matches
+				fn := v.Type.(*btf.Func)
+				if fn.Name != name {
+					continue
+				}
+				ins.Metadata.Set(kfuncMeta{}, fn)
+				ins.Src = asm.PseudoKfuncCall
+				ins.Constant = -1
+
+				found = true
+				break
+			}
+
+			if !found {
+				return fmt.Errorf("kfunc %s not found in %s", rel.Name, ".ksyms")
```

Comment:
I think this shouldn't be an error, otherwise we break the non `.ksyms` extern hack. We should just fall through to the end of the loop.